families/block_info/sawtooth_block_info/processor/handler.py

```diff
@@ -0,0 +1,179 @@
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+import logging
+import time
+
+from sawtooth_sdk.processor.state import StateEntry
+from sawtooth_sdk.processor.exceptions import InvalidTransaction
+from sawtooth_sdk.processor.exceptions import InternalError
+
+from sawtooth_block_info.protobuf.block_info_pb2 import BlockInfoTxn
+from sawtooth_block_info.protobuf.block_info_pb2 import BlockInfo
+from sawtooth_block_info.protobuf.block_info_pb2 import BlockInfoConfig
+
+from sawtooth_block_info.common import FAMILY_NAME
+from sawtooth_block_info.common import FAMILY_VERSION
+from sawtooth_block_info.common import ENCODING
+from sawtooth_block_info.common import NAMESPACE
+from sawtooth_block_info.common import CONFIG_ADDRESS
+from sawtooth_block_info.common import DEFAULT_SYNC_TOLERANCE
+from sawtooth_block_info.common import DEFAULT_TARGET_COUNT
+from sawtooth_block_info.common import create_block_address
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class BlockInfoTransactionHandler(object):
+    @property
+    def family_name(self):
+        return FAMILY_NAME
+
+    @property
+    def family_versions(self):
+        return [FAMILY_VERSION]
+
+    @property
+    def encodings(self):
+        return [ENCODING]
+
+    @property
+    def namespaces(self):
+        return [NAMESPACE]
+
+    def apply(self, transaction, state):
+        # Unpack payload
+        txn = BlockInfoTxn()
+        txn.ParseFromString(transaction.payload)
+
+        # Validate block info included in payload
+        next_block = txn.block
+        if any([next_block.block_num < 0,
+                next_block.previous_block_id == "",
+                next_block.signer_pubkey == "",
+                next_block.header_signature == "",
+                next_block.timestamp < 0]):
+            raise InvalidTransaction(
+                "All fields in BlockInfo must be set")
+
+        # Get config and previous block (according to the block info in the
+        # transaction) from state
+        entries = state.get([CONFIG_ADDRESS])
+
+        deletes = []
+        sets = []
+        config = BlockInfoConfig()
+
+        # If there is no config in state, we don't know anything about what's
+        # in state, so we have to treat this as the first entry
+        if not entries:
+            # If sync tolerance or target count were not specified in the txn,
+            # use default values.
+            config.sync_tolerance = \
+                DEFAULT_SYNC_TOLERANCE if txn.sync_tolerance == 0 \
+                else txn.sync_tolerance
+
+            config.target_count = \
+                DEFAULT_TARGET_COUNT if txn.target_count == 0 \
+                else txn.target_count
+
+            config.latest_block = next_block.block_num
+            config.oldest_block = next_block.block_num
+
+            now = time.time()
+            if next_block.timestamp > now + config.sync_tolerance:
```

Comment:
This should be `abs(now - next_block.timestamp) > config.sync_tolerance` to account for the peer being slightly behind